standalone-cinder: Code organization and unit tests

aglitke opened the pull request:
This PR reorganizes the standalone-cinder code into packages (provisioner and volumeservice) and creates a separate cmd that consumes them.  The initial submission was missing unit tests.  In order to add them, the interfaces between components (provisioner, Volume Mappers, and Volume Service) had to be cleaned up.  In order to enable unit tests some broker interfaces have been added to allow mock implementations to be substituted during testing.  Finally, unit tests are added for the provisioner package covering just about all of the code.

rootfs commented:
@aglitke please rebase

rootfs commented on openstack/standalone-cinder/pkg/provisioner/provisioner.go:
```diff
@@ -0,0 +1,176 @@
+
+	err = p.vsb.WaitForAvailableCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Volume did not become available")
```
also print volumeID

rootfs commented on openstack/standalone-cinder/pkg/provisioner/provisioner.go:
```diff
@@ -0,0 +1,176 @@
+	err = p.vsb.ReserveCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to reserve volume: %v", err)
```
volumeID here (and below) too

rootfs commented on openstack/standalone-cinder/pkg/provisioner/provisioner.go:
```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to reserve volume: %v", err)
+		return nil, err
```
clean up the volume upon error (and below too)

aglitke commented:
@rootfs I need ginkgo and gomega for my testing code.  It is already being used privately by the nfs tests.  Do you think there would be support to pull them into the main glide.yaml file?

rootfs commented:
@wongma7 for that call

wongma7 commented:
Yes it's fine to add ginkgo and gomega. The reason the nfs tests have their own dependencies is they import the kubernetes e2e framework and I needed to keep that dependency locked independent of the rest of the repo

aglitke commented:
@rootfs I've addressed your comments in the current commits.  I now rollback all provisioning operations when a failure occurs.  I'm also doing a better job printing the cinder volume ID when cinder calls fail.

rootfs commented on openstack/standalone-cinder/pkg/provisioner/iscsi.go:
```diff
@@ -0,0 +1,92 @@
+	secretName := getChapSecretName(conn, options)
+	if secretName != "" {
+		ret.ISCSI.SessionCHAPAuth = true
+		ret.ISCSI.SecretRef = &v1.LocalObjectReference{
```
In 1.8, this changes to SecretReference. Leave a note here.

aglitke replied:
Ok. What is the k8s policy on backwards-compatibility?  Can I expect this method to work for one or two minor versions?

rootfs replied:
whenever client-go is upgraded to 1.8 :)

rootfs commented on openstack/standalone-cinder/pkg/provisioner/provisioner.go:
```diff
@@ -0,0 +1,201 @@
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
```
check access mode, see https://github.com/kubernetes-incubator/external-storage/blob/master/gluster/block/cmd/glusterblock-provisioner/glusterblock-provisioner.go#L168

rootfs replied:
Sure, leave a TODO here

aglitke commented on openstack/standalone-cinder/pkg/provisioner/iscsi.go:
```diff
@@ -0,0 +1,92 @@
+	ret := &v1.PersistentVolumeSource{
+		ISCSI: &v1.ISCSIVolumeSource{
+			// TODO: Need some way to specify the initiator name
+			TargetPortal:    conn.Data.TargetPortal,
```
Not yet.  This will definitely be something to add in the future.

rootfs commented on openstack/standalone-cinder/pkg/provisioner/provisioner.go:
```diff
@@ -0,0 +1,201 @@
+ERROR_DISCONNECT:
+	cleanupErr = p.vsb.DisconnectCinderVolume(p.VolumeService, volumeID)
+	if cleanupErr != nil {
+		glog.Errorf("Failed to disconnect volume %s: %v", volumeID, cleanupErr)
```
I don't know the best answer but later in `Delete()`, similar error leaves rest of the code unexecuted. Some consistency is preferred

aglitke replied:
I am not sure the best answer here either.  If something is wrong with cinder or the storage backend the cleanup actions aren't too likely to succeed either.  I've been deferring a decision on how to handle this until I see more actual scenarios in the wild.  I can change the Delete flow to work the same as the Provision cleanup path if that makes more sense.

rootfs commented on openstack/standalone-cinder/pkg/provisioner/vsbroker.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+type volumeServiceBroker interface {
+	CreateCinderVolume(vs *gophercloud.ServiceClient, options controller.VolumeOptions) (string, error)
```
nit: godoc

aglitke replied:
This is an interface used just for overriding cinder calls in the volumeservice module.  I'll add a comment about the interface at the top.

rootfs commented on openstack/standalone-cinder/pkg/provisioner/mapper.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+type volumeMapper interface {
+	BuildPVSource(conn volumeservice.VolumeConnection, options controller.VolumeOptions) (*v1.PersistentVolumeSource, error)
```
nit: godoc

aglitke replied:
Done.

aglitke commented on openstack/standalone-cinder/pkg/provisioner/provisioner.go:
```diff
@@ -0,0 +1,201 @@
+	err = p.vsb.WaitForAvailableCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Volume %s did not become available", volumeID)
+		goto ERROR_DELETE
```
This comes from my days as a Linux kernel programmer :smile:  I think goto (when used properly) can produce very readable error cleanup code.  Since I am curious, would you implement a cleanup function like this:
```go
func cleanup(unreserve bool, disconnect bool, delete bool) {
    if unreserve {
        // unreserve volume
    }
    if disconnect {
        // disconnect volume
    }
    if delete {
        // delete volume
    }
}
```

rootfs replied:
sgtm

rootfs commented:
/lgtm
